### ISSUE TITLE
as a senior editor I want to mark a task as 'blocked' until another task has been done

### ISSUE THREAD

**Issue by goose-life** [laws-africa/indigo#1140]
e.g. on https://edit.laws.africa/works/akn/za/act/1957/33/, 
the first amendment task is blocked by the import task,
the second amendment task is blocked by the first,
etc.

### Suggestion
- Add a 'blocked by' field on a task, that takes one or more task numbers as its input
- Disable such a task from being picked up and automatically add a `blocked` label
- Once the blocking tasks have been done, make the task pick up able and remove the label

**Comment by goose-life** [comment]
The only examples we've had so far of editors wanting to 'block' tasks are for applying amendments until the previous point in time has been published.

## Proposal 1:
- Add 'blocked' field on the Task model, which we can model richly or not (see above) 
- Users can't assign or pick up a blocked task 

If we model task types, we may be able to lean on this to, say, include 'blocked by Task X' only on 'Apply amendment' tasks.

## Proposal 2:
Block a new point in time from being created on a Work if 
- A previous point in time exists? (This would block a consolidation document from being imported at the latest point in time, so probably not a great idea) 
- A document exists at a previous point in time but hasn't been published (This won't block someone from creating a document at the latest point in time and attempting to 'Apply amendments' to a non-existent document. It wouldn't make a lot of sense for someone to do this assuming they have the necessary context, but we need to keep in mind that picking up a task from 'All available tasks' doesn't always facilitate that.)

My vote is for Proposal 1. We've considered modelling task types more strongly for other good reasons and I think the challenge we're looking to address will be better served by the task system. It'll allow us to automatically share all the context needed to do a task without overloading a user with TMI, while allowing users to technically do the 'wrong' thing (e.g. importing a consolidation) without needing to special-case it.

@longhotsummer, thoughts?

**Comment by goose-life** [comment]
## Proposal 1 fleshed out a bit:

- Add `blocked_by` to Task (m2m, because a task can both block and be blocked by more than one other task)

Task form:
- New 'Blocked by other task(s):' with multi select dropdown of all other tasks on this work (note: this task might not have a pk yet)
- (If the list of potential blocking tasks needs to be expanded in future we can do that, but this should suffice for our current needs)
- Automatically add the `BLOCKED` label (if it exists) on saving (We still want them to appear in Available tasks for insight, but we don't want them to be clicked through to.) (This could probably be handled more intelligently.)

Bulk import:
~When~After creating an 'apply-amendment' task on a work:
- Check that work's existing tasks for `apply-amendment` or `import-content` tasks (and stash them)
- If they exist, add them to the task's `blocked_by` list, and add the `BLOCKED` label (if it exists)
- The task will need to be saved first because of the M2M relation, so do this way at the end; that way it also doesn't matter what order the works (and their tasks) were created in

Disable all 'Assign' buttons if a task has a non-empty `blocked_by` list:
- Single task
- Task list
- Workflow
- ??

Whenever any task is closed:
- check if it's blocking another task
- remove it from that task's `blocked_by` list
- if that task's `blocked_by` list is now empty, remove the `BLOCKED` label (if it exists and it's on the task)

**Comment by longhotsummer** [comment]
Proposal 1 is great. We have had times when we haven't imported from a spreadsheet because we didn't want to do allow the tasks to be done. With this, we could import and then mark them blocked. Maybe we make "blocked" a task state? That way we get the state transition protection, can filter on it using existing filters, etc.

My concern with proposal 2 is it's too smart, and would be difficult to work around if we felt we knew better. Instead, I think we should do proposal 1 AND add a warning notice when someone is about to create a new point in time from an older point in time that doesn't look "complete" (ie. published, as you suggest).

**Comment by longhotsummer** [comment]
The 'blocked' state may also remove the need for a 'blocked' label.

**Comment by goose-life** [comment]
I love having `blocked` as a state, thanks!

**Comment by goose-life** [comment]
## Final Proposal 1:

- Add `blocked_by` to Task (m2m, because a task can both block and be blocked by more than one other task)
- Add `blocked` as a Task state in CLOSED_STATES
- Add block_task permission and add it to the Reviewer badge
- Add icon for blocked tasks (I'm thinking a red ×)
- Add optional `date` field to Task, which will be the point in time at which it needs to be done

Case for having it in OPEN_STATES
- It'll still show up in Available tasks, but the icon will show that it's currently blocked

Case for having it in CLOSED_STATES
- It won't be assignable
- We can still show blocked tasks by default in task listing views, and show them on the work overview under 'Open tasks'

Therefore:
- have it in CLOSED_STATES
- show it in Tasks and Work overview by default, but not Available tasks (because it's not available, duh)

Task detail:
- Add same dropdown as for workflows, but for 'Blocked by task(s)', only visible to Reviewers
- If a user selects one or more blocked_by tasks, move it into `blocked` by that user
- If it's already blocked and a user empties out the blocked_by list, move it into `open` by that user
- Hide the Reopen button when a task is in `blocked` state

Task edit form:
- add `date` dropdown which has all the points in time on a work as options

Bulk import:
- Add the relevant date to each Import and Apply amendment task
- After creating all works, go through all the 'apply-amendment' tasks that were created:
- Check other tasks on the associated work for `import-content` or `apply-amendment` (before this task's date) tasks (and stash them)
- If they exist, add them to the task's `blocked_by` list
- Move the task into the `blocked` state

Whenever **any** task is closed:
- clear its `blocking` list
- for each task that was in the `blocking` list, if that task's `blocked_by` list is now empty, move the task into `open`

If a blocking task is reopened, the other tasks will need to be manually blocked by that task again.

**Comment by goose-life** [comment]
### Extra thoughts

1. Also make a task blockable from the Edit task form (same as 'Cancel'), without having to choose a blocked_by task? We could for example move all `File missing`, `Gazette missing`, and `Not yet supported` tasks into `blocked`.
Pros:
- Sometimes a task is blocked by something other than another task, and we want to capture that.
Cons:
- If a task is in fact blocked by another task, but it was easier to simply click 'Block' than to choose the right blocking task from the dropdown, when that other task is eventually resolved this task won't automatically be moved into `open`.

So there should be the option, but it should be easier to choose a blocking task than to simply 'Block' a task.
In other words:
- New 'Block' button in the edit form (next to 'Cancel').
- 'Reopen' button on task detail view, disabled as long as a task has a non-empty `blocked_by` list but enabled if a task is simply `blocked`.

2. Task bulk update form should allow one to bulk block tasks (iirc currently one can only bulk assign them). It should also be possible to filter to tasks and 'select all' for bulk update.

3. If the Gazette of an amending work is missing, we should choose the relevant 'Link Gazette' task , but that'll be on the other work. So instead of including all tasks on this work, include all tasks on this _and related_ works in the 'Blocked by' dropdown. They'll need to be grouped by work to be useful.

**Comment by longhotsummer** [comment]
> Add blocked as a Task state in CLOSED_STATES

I'm not convinced by this. CLOSED_STATES is used elsewhere in the code and has a specific meaning. A blocked task is still open. I'd rather we put it into OPEN_STATES and then review where OPEN_STATES is used and ensure it continues to makes sense, and adjust accordingly.

> Hide the Reopen button when a task is in blocked state

This would then be unnecessary.

A mockup on Balsamiq of the various buttons would be useful.

> Also make a task blockable from the Edit task form (same as 'Cancel'), without having to choose a blocked_by task? We could for example move all File missing, Gazette missing, and Not yet supported tasks into blocked.

I had a similar thought and agree with this idea. 

> Add same dropdown as for workflows, but for 'Blocked by task(s)', only visible to Reviewers

I think if a user can edit a task, they should be able to mark it as blocked or unblocked. The blocked state, and what it's blocked on, should be visible to anyone who can see the task.

Task Detail view should also show what tasks this task blocks (must click through to blocked task to edit).

**Comment by longhotsummer** [comment]
Apart from the above, this looks good to me and we can start implementing.

**Comment by goose-life** [comment]
## Final Proposal 1 (again):

- Add `blocked_by` to Task (m2m, because a task can both block and be blocked by more than one other task)
- Add `blocked` as a Task state in OPEN_STATES
- Add block_task permission and add it to the Editor badge
- Add icon for blocked tasks (I'm thinking a red ×)
- Add optional `date` field to Task, which will be the point in time at which it needs to be done

Task listing views:
- Show blocked tasks by default, along with their icon so that users know they won't be able to pick them up
- Potentially hide blocked tasks from 'All available tasks' view (TBD)

Task detail:
- List tasks that this task is blocking, if any
- Add same dropdown as for projects, but for 'Blocked by task(s)'
- The potential blocking tasks listed here should be all tasks on this work, as well as all tasks on related works, grouped by work
- If a user selects one or more blocked_by tasks, move it into `blocked` by that user
- If it's blocked and a user empties out the blocked_by list, move it into `open` by that user

Task edit form:
- Add `date` dropdown which has all the points in time on a work as options
- New 'Block' button in the edit form (next to 'Cancel')
- New 'Unblock' button when a task is in `blocked` state and has a non-empty blocked_by list (next to 'Cancel')

Bulk import:
- Add the relevant date to each Import and Apply amendment task
- After creating all works, go through all the 'apply-amendment' tasks that were created:
- Check other tasks on the associated work for `import-content` or `apply-amendment` (before this task's date) tasks (and stash them)
- If they exist, add them to the task's `blocked_by` list
- Move the task into the `blocked` state, save and record the relevant activity-stream actions

Whenever **any** task is closed:
- clear its `blocking` list
- for each task that was in the `blocking` list, if that task's `blocked_by` list is now empty, move the task into `open`, and record the relevant activity-stream actions

Task bulk update form:
- Allow one to bulk block tasks (or change their state some other way)
- Allow one to filter to tasks and 'select all' for bulk update

If a blocking task is reopened, the other tasks will need to be manually blocked by that task again.

**Comment by goose-life** [comment]
![image](https://user-images.githubusercontent.com/32566441/106713193-c40baf80-6602-11eb-8e91-79f0c6c83883.png)

**Comment by goose-life** [comment]
![image](https://user-images.githubusercontent.com/32566441/106713226-d1c13500-6602-11eb-92fe-fe281a8e0f67.png)

**Comment by longhotsummer** [comment]
Nice. Some thoughts.

1. The blocked state needs to be clearer. How about an alert block `This task is blocked and cannot be done [until these tasks are completed: ...]` and a list of the blocked tasks, right at the top above the task description. 
2. Put the tasks this tasks is blocking right at the bottom of the sidebar, change the heading to `Tasks blocked by this task` (whenever you find yourself using "the following" or "the items below" or similar in UI text, you know you need to re-phrase it)
3. When a task is blocked, it must be unassigned (otherwise we need to be able to assign/unassign a blocked task).
4. I'm not sure about the date -- can we leave that out of this update? eg. how do we communicate that this is not a date by which to do this task (which is what I thought it was originally)?
5. What is the experience when choosing new blocking tasks? Is that a dropdown (there can be 1000s of open tasks in a place), or is it a popup like when choosing a work? Either way, they will need to be searchable by title and/or number.
4. We should really make labels editable into the normal task view, but let's not do that now.

**Comment by longhotsummer** [comment]
Let's try this for a dropdown, and if it works we can duplicate it for the projects and potentially labels, too.

The "Blocked by" header is actually a bootstrap-select widget, with fixed text. We load all the open tasks by ajax when the user clicks on that dropdown for the first time. This avoids us having to do server-side search, and saves us from loading 1000+ open tasks every time that page loads. Robley or I can assist with this.
